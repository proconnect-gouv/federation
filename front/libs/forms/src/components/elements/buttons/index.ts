export * from './array-add';
export * from './array-remove';
export * from './clipboard';
