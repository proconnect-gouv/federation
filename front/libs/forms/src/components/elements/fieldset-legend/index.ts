export * from './fieldset-legend.element';
