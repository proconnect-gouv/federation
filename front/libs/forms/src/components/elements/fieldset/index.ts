/* istanbul ignore file */

// declarative file
export * from './fieldset.element';
