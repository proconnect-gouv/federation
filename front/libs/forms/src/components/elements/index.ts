export * from './fieldset';
export * from './fieldset-legend';
export * from './group';
export * from './label';
export * from './message';
