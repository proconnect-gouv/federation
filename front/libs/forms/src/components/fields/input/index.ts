/* istanbul ignore file */

// declarative file
export * from './input.field';
