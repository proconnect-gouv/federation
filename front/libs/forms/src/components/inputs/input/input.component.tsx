import React from 'react';
import type { FieldInputProps } from 'react-final-form';

import type { PropsWithClassName } from '@fc/common';

interface InputComponentProps<FieldValue = string> extends PropsWithClassName {
  readOnly?: boolean;
  disabled?: boolean;
  id: string;
  input: FieldInputProps<FieldValue, HTMLElement | HTMLSelectElement>;
}

export const InputComponent = React.memo(
  ({ className, disabled = false, id, input, readOnly = false }: InputComponentProps) => (
    <input
      {...input}
      aria-describedby={`${input.name}-messages`}
      className={className}
      data-testid={`${id}--testid`}
      disabled={disabled}
      id={id}
      name={input.name}
      readOnly={readOnly}
      type={input.type}
    />
  ),
);

InputComponent.displayName = 'InputComponent';
