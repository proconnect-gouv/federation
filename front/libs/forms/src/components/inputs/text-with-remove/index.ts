export * from './text-with-remove.input';
