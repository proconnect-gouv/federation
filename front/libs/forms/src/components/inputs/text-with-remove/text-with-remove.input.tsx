import type { FieldValidator } from 'final-form';
import React, { useCallback } from 'react';
import { useField } from 'react-final-form';

import { ComponentTypes } from '../../../enums';
import { useFieldMeta } from '../../../hooks';
import { ArrayRemoveButton, GroupElement, MessageElement } from '../../elements';
import { InputComponent } from '../input';

interface TextWithRemoveInputProps {
  index: number;
  fieldName: string;
  isRemovable?: boolean;
  onRemove: (index: number) => void;
  validate: FieldValidator<string> | undefined;
}

export const TextWithRemoveInput = React.memo(
  ({ fieldName, index, isRemovable = false, onRemove, validate }: TextWithRemoveInputProps) => {
    const { input, meta } = useField(fieldName, { validate });
    const { errorMessage, hasError, inputClassname, isValid } = useFieldMeta(meta);

    const removeHandler = useCallback(() => {
      onRemove(index);
    }, [index, onRemove]);

    const { className, disabled, name } = input;
    const id = `form-input-array-${name}`;
    return (
      <GroupElement
        className={className}
        disabled={disabled}
        hasError={hasError}
        isValid={isValid}
        type={ComponentTypes.INPUT}>
        <div className="flex-columns">
          <InputComponent className={inputClassname} disabled={false} id={id} input={input} />
          <ArrayRemoveButton
            dataTestId={`${fieldName}-remove`}
            disabled={!isRemovable}
            onClick={removeHandler}
          />
        </div>
        <MessageElement
          dataTestId={`${name}-messages`}
          error={errorMessage}
          id={id}
          isValid={isValid}
        />
      </GroupElement>
    );
  },
);

TextWithRemoveInput.displayName = 'TextWithRemoveInput';
