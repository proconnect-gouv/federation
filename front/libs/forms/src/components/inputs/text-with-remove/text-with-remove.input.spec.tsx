import { fireEvent, render } from '@testing-library/react';
import React from 'react';
import { type FieldInputProps, useField } from 'react-final-form';

import { useFieldMeta } from '../../../hooks';
import { ArrayRemoveButton, GroupElement, MessageElement } from '../../elements';
import { InputComponent } from '../input';
import { TextWithRemoveInput } from './text-with-remove.input';

jest.mock('../input/input.component');
jest.mock('../../elements/group/group.element');
jest.mock('../../elements/label/label.element');
jest.mock('../../elements/message/message.element');
jest.mock('../../elements/buttons/array-remove/array-remove.button');
jest.mock('../../../hooks/field-meta/field-meta.hook');

describe('TextWithRemoveInput', () => {
  // Given
  const isValidMock = Symbol('is-valid-mock') as unknown as boolean;
  const disabledMock = Symbol('disabled-mock') as unknown as boolean;
  const hasErrorMock = Symbol('has-error-mock') as unknown as boolean;

  const metaMock = {
    error: 'any-errorMessage-mock',
    touched: false,
    value: 'any-input-value-mock',
  };

  const inputMock = {
    className: 'any-classname-mock',
    disabled: disabledMock,
    name: 'any-name-mock',
    type: 'text',
    value: 'any-input-value-mock',
  } as unknown as FieldInputProps<string, HTMLElement>;

  beforeEach(() => {
    // Given
    jest.mocked(useField).mockReturnValue({
      input: inputMock,
      meta: metaMock,
    });
    jest.mocked(useFieldMeta).mockReturnValue({
      errorMessage: 'any-error-message-mock',
      hasError: hasErrorMock,
      inputClassname: 'any-input-classname-mock',
      isValid: isValidMock,
    });
  });

  it('should match snapshot', () => {
    // Given
    const validateMock = jest.fn();
    const onRemoveMock = jest.fn();
    const isRemovableMock = false;

    const useCallbackMock = jest.fn();
    jest.spyOn(React, 'useCallback').mockImplementationOnce(() => useCallbackMock);

    // When
    const { container } = render(
      <TextWithRemoveInput
        fieldName="any-fieldname-mock"
        index={0}
        isRemovable={isRemovableMock}
        validate={validateMock}
        onRemove={onRemoveMock}
      />,
    );

    // Then
    expect(container).toMatchSnapshot();
    expect(useFieldMeta).toHaveBeenCalledOnce();
    expect(useFieldMeta).toHaveBeenCalledWith(metaMock);
    expect(GroupElement).toHaveBeenCalledOnce();
    expect(GroupElement).toHaveBeenCalledWith(
      {
        children: expect.any(Object),
        className: 'any-classname-mock',
        disabled: disabledMock,
        hasError: hasErrorMock,
        isValid: isValidMock,
        type: 'input',
      },
      {},
    );
    expect(InputComponent).toHaveBeenCalledOnce();
    expect(InputComponent).toHaveBeenCalledWith(
      {
        className: 'any-input-classname-mock',
        disabled: false,
        id: 'form-input-array-any-name-mock',
        input: inputMock,
      },
      {},
    );
    expect(ArrayRemoveButton).toHaveBeenCalledOnce();
    expect(ArrayRemoveButton).toHaveBeenCalledWith(
      {
        dataTestId: 'any-fieldname-mock-remove',
        disabled: !isRemovableMock,
        onClick: useCallbackMock,
      },
      {},
    );
    expect(MessageElement).toHaveBeenCalledOnce();
    expect(MessageElement).toHaveBeenCalledWith(
      {
        dataTestId: 'any-name-mock-messages',
        error: 'any-error-message-mock',
        id: 'form-input-array-any-name-mock',
        isValid: isValidMock,
      },
      {},
    );
  });

  it('should call onRemove with index when user click the remove button', () => {
    // Given
    const onRemoveMock = jest.fn();
    const indexMock = Symbol('index-mock') as unknown as number;

    jest.mocked(ArrayRemoveButton).mockImplementationOnce(({ onClick }) => (
      <button type="button" onClick={onClick}>
        any-remove-button-mock
      </button>
    ));

    // When
    const { getByText } = render(
      <TextWithRemoveInput
        fieldName="any-fieldname-mock"
        index={indexMock}
        validate={jest.fn()}
        onRemove={onRemoveMock}
      />,
    );
    const buttonElt = getByText('any-remove-button-mock');
    fireEvent.click(buttonElt);

    // Then
    expect(onRemoveMock).toHaveBeenCalledOnce();
    expect(onRemoveMock).toHaveBeenCalledWith(indexMock);
  });
});
