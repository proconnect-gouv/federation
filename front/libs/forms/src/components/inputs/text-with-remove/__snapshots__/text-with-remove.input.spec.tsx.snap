// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextWithRemoveInput should match snapshot 1`] = `
<div>
  <div
    data-mockid="GroupElement"
  >
    <div
      class="flex-columns"
    >
      <div
        data-mockid="InputComponent"
      />
      <div
        data-mockid="ArrayRemoveButton"
      />
    </div>
    <div
      data-mockid="MessageElement"
    />
  </div>
</div>
`;
